test/e2e/metrics_e2e_test.go

```diff
@@ -266,6 +271,110 @@ var _ = Describe("Metrics are generated for OLM managed resources", func() {
 			})
 		})
 	})
+
+	Context("Metrics emitted by CatalogSources", func() {
+		When("A valid CatalogSource object is created", func() {
+			var (
+				name    = "metrics-catsrc-valid"
+				cleanup func()
+			)
+			BeforeEach(func() {
+				mainPackageName := genName("nginx-")
+
+				mainPackageStable := fmt.Sprintf("%s-stable", mainPackageName)
+
+				stableChannel := "stable"
+
+				mainCRD := newCRD(genName("ins-"))
+				mainCSV := newCSV(mainPackageStable, testNamespace, "", semver.MustParse("0.1.0"), []apiextensions.CustomResourceDefinition{mainCRD}, nil, nil)
+
+				mainManifests := []registry.PackageManifest{
+					{
+						PackageName: mainPackageName,
+						Channels: []registry.PackageChannel{
+							{Name: stableChannel, CurrentCSVName: mainPackageStable},
+						},
+						DefaultChannelName: stableChannel,
+					},
+				}
+				_, cleanupAll := createInternalCatalogSource(c, crc, name, testNamespace, mainManifests, []apiextensions.CustomResourceDefinition{mainCRD}, []v1alpha1.ClusterServiceVersion{mainCSV})
+
+				var once sync.Once
+				cleanup = func() {
+					once.Do(cleanupAll)
+				}
+			})
+			AfterEach(func() {
+				cleanup()
+			})
+			It("emits metrics for the catalogSource", func() {
+				Eventually(func() []Metric {
+					return getMetricsFromPod(c, getPodWithLabel(c, "app=catalog-operator"), "8081")
+				}).Should(And(
+					ContainElement(LikeMetric(
+						WithFamily("catalog_source_count"),
+						WithValueGreaterThan(0),
+					)),
+					ContainElement(LikeMetric(
+						WithFamily("catalogsource_ready"),
+						WithName(name),
+						WithNamespace(testNamespace),
+						WithValue(1),
+					)),
+				))
+			})
+			When("The CatalogSource object is deleted", func() {
+				BeforeEach(func() {
+					cleanup()
+				})
+				It("deletes the metrics for the CatalogSource", func() {
+					Eventually(func() []Metric {
+						return getMetricsFromPod(c, getPodWithLabel(c, "app=catalog-operator"), "8081")
+					}).Should(And(
+						Not(ContainElement(LikeMetric(
+							WithFamily("catalogsource_ready"),
+							WithName(name),
+							WithNamespace(testNamespace),
+						)))))
+				})
+			})
+		})
+
+		When("A CatalogSource object is in an invalid state", func() {
+			var (
+				name    = "metrics-catsrc-invalid"
+				cleanup func()
+			)
+			BeforeEach(func() {
+				_, cleanup = createInvalidGRPCCatalogSource(crc, name, testNamespace)
+			})
+			AfterEach(func() {
+				cleanup()
+			})
+			It("emits metrics for the CatlogSource with a Value greater than 0", func() {
+				Eventually(func() []Metric {
+					return getMetricsFromPod(c, getPodWithLabel(c, "app=catalog-operator"), "8081")
+				}).Should(And(
+					ContainElement(LikeMetric(
+						WithFamily("catalogsource_ready"),
+						WithName(name),
+						WithNamespace(testNamespace),
+						WithValue(0),
+					)),
+				))
+				Consistently(func() []Metric {
+					return getMetricsFromPod(c, getPodWithLabel(c, "app=catalog-operator"), "8081")
+				}, "3m").Should(And(
```

Comment:
@anik120 Any idea on why "3m" was chosen as the consistently poll here? I was poking around this package trying to rundown some flakes and noticed specific test spec takes roughly 75-80% of the total runtime for this metrics e2e package.

Reply:
@timflannagan I can't seem to remember why I chose 3m, but I think I wanted to prove out that the metrics with a particular value is emitted for a considerable amount of time. Looking back at it, I don't think `Eventually` with a `pollAfter` duration set is a bad idea either. That'll reduce the run time since I'm assuming the process will sleep till it's time to poll.

Reply:
Actually, there are rules that depend on this metric being having a consistent value over a period of time, eg https://github.com/operator-framework/operator-marketplace/blob/master/manifests/12_prometheus_rule.yaml#L23-L24. We could reduce the value to 2m too if that'll help bring that percentage down to ~50%

Reply:
Opened a PR that's a middle ground solution (I think) https://github.com/operator-framework/operator-lifecycle-manager/pull/2739